File: JSOI_2010_GROUP_部落划分_GROUP.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long

struct DSU { // 并查集
    vector<int> pre, siz;
    
    DSU() {}
    DSU(int n) {
        pre.resize(n + 1);
        iota(pre.begin(), pre.end(), 0);
        siz.assign(n + 1, 1);
    }
    
    int find(int x) {
        if (pre[x] == x) {
            return x;
        }
        return pre[x] = find(pre[x]);
    }
    
    bool same(int x, int y) {
        return find(x) == find(y);
    }
    
    bool merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) {
            return false;
        } else if (siz[x] < siz[y]) {
            swap(x, y);
        }
        siz[x] += siz[y];
        pre[y] = x;
        return true;
    }
    
    int size(int x) {
        return siz[find(x)];
    }
};

signed main() {
    int n, k, cnt = 0;
    cin >> n >> k;
    DSU dsu(n);
    vector<pair<int, int>> points(n);
    vector<tuple<double, int, int>> dis;
    for (int i = 0; i < n; ++i) {
        cin >> points[i].first >> points[i].second;
        for (int j = 0; j < i; ++j) {
            int dx = points[i].first - points[j].first;
            int dy = points[i].second - points[j].second;
            double dist = sqrt(dx * dx + dy * dy);
            dis.emplace_back(dist, i, j);
        }
    }
    sort(dis.begin(), dis.end());
    for (auto [l, x, y] : dis) {
        if (dsu.merge(x, y)) {
            // cnt++;
            if (++cnt == n - k + 1) {
                cout << fixed << setprecision(2) << l << endl;
                return 0;
            }
        }
    }
}

File: JSOI_2010_GROUP_部落划分_GROUP.py
import sys
input = lambda: sys.stdin.readline().strip()
from math import sqrt

class DSU: # 并查集
    def __init__(self, n):
        self.pre = list(range(n + 1))
        self.siz = [1 for _ in range(n + 1)]
    
    def find(self, x):
        x, xcopy = self.pre[x], x
        while x != self.pre[x]:
            x = self.pre[x]
        while xcopy != x:
            self.pre[xcopy], xcopy = x, self.pre[xcopy]
        return x
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def merge(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        self.siz[x] += self.siz[y]
        self.pre[y] = x
        return True
    
    def size(self, x):
        return self.siz[self.find(x)]

n, k = map(int, input().split())
dsu = DSU(n)
cnt = 0
points, dis = [], []
points = [list(map(int, input().split())) for _ in range(n)]
for i in range(n):
    x, y = points[i]
    for j in range(i):
        xx, yy = points[j]
        dx, dy = x - xx, y - yy
        dist = sqrt(dx * dx + dy * dy)
        dis.append((dist, i, j))
dis.sort()
for l, x, y in dis:
    if dsu.merge(x, y):
        cnt += 1
        if cnt == n - k + 1:
            print(f"{l:.2f}")
